packit_service/models.py

```diff
@@ -2329,6 +2329,37 @@ def add_scan(self, task_id: int) -> "OSHScanModel":
             session.add(scan)
             return scan
 
+    @contextmanager
+    def add_scan_transaction(self) -> Generator["OSHScanModel"]:
+        """
+        Context manager that creates a ScanModel upon entering the context,
+        provides a corresponding instance of `ScanModel` to be updated within the context
+        and commits the changes upon exiting the context, all within a single transaction.
+        """
+        with sa_session_transaction(commit=True) as session:
+            # lock self in the DB - all other workers will block here until the context is exited
+            session.query(type(self)).filter_by(id=self.id).with_for_update().first()
+
+            new_scan = None
+            try:
+                if not self.scan:
+                    new_scan = OSHScanModel()
+                    new_scan.copr_build_target = self
+                    session.add(new_scan)
+            except Exception as ex:
+                logger.warning(f"Exception while working with database: {ex!r}")
+                session.rollback()
+                raise
+
+            try:
+                if new_scan:
+                    yield new_scan
+                    session.add(new_scan)
+                    session.commit()
```

Comment:
`sa_session_transaction(commit=True)` does the commit for you. As well as the rollback above and below. The only issue is that a potential exception will be prefixed with a DB exception text even if it's not DB related, not sure how to elegantly avoid that.